File: utils/c/input.c
#include "input.h"
#include "helper.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

char *getinput(const char *restrict path) {
  FILE *f = fopen(path, "rb");

  fseek(f, 0, SEEK_END);
  long fsize = ftell(f);
  fseek(f, 0, SEEK_SET);

  char *string = malloc((fsize + 1) * sizeof(char));
  fread(string, fsize, 1, f);
  fclose(f);

  string[fsize] = 0;
  // yeet empty last line that always somehow happen for no reason
  // i hope this will totally not screw me over
  if (string[fsize - 1] == '\n') {
    char *cpy = malloc(fsize * sizeof(char));
    strncpy(cpy, string, fsize - 1);
    cpy[fsize - 1] = 0;
    string = cpy;
  }

  return string;
}

Answers *getanswers(const char *restrict path) {
  int sz, line = 0;
  char **input = strsplit(getinput(path), "\n", &sz);
  Answers *answers = malloc(sizeof(Answers));

  while (line < sz) {
    switch (line) {
    case 0:
      answers->test1 = input[line];
      break;
    case 1:
      answers->part1 = input[line];
      break;
    case 2:
      answers->test2 = input[line];
      break;
    case 3:
      answers->part2 = input[line];
      break;
    }
    line++;
  }

  return answers;
}

File: utils/ts/input.ts
export type Answers = {
   part1: string;
   part2: string;
   test1: string;
   test2: string;
};

/**
 * Return sanitised input and removes redundant line break at the end.
 * @param path
 * @returns
 */
export function getInput(path: string): string {
   return Deno.readTextFileSync(path).replaceAll('\r', '').replace(/\n$/, '');
}

/**
 * Obtain answers from `answers.txt` file.
 * @param path pass `import.url.meta`, must be in a child directory as parent directory will contain the answers file.
 * @returns {Answers} answers object
 */
export function getAnswers(path: string): Answers {
   return getInput(path)
      .split('\n')
      .reduce(
         (answers, value, idx) => {
            value = value.trim();
            switch (idx) {
               case 0:
                  answers.test1 = value;
                  break;
               case 1:
                  answers.part1 = value;
                  break;
               case 2:
                  answers.test2 = value;
                  break;
               case 3:
                  answers.part2 = value;
                  break;
            }
            return answers;
         },
         {
            part1: '',
            part2: '',
            test1: '',
            test2: '',
         } satisfies Answers
      );
}
